### PR TITLE
resolves date range issue #31

### PR DESCRIPTION
**Pull request by mshodge**
Resolves reopened issue #31 

Have also added ability to choose month_from and month_to. The former chooses the first day of that month and the latter the last day of the month. So individual months can be run using the same MM for both. Also a few args checks to make sure year is in YYYY format and month MM, and the month_from and year_from cannot be in the future of the month_to and year_to.

year2pandas_latest_date and year2pandas_earliest_date functions could be optimised as I wrote them using lengthy versions of is statements.

Example run:

python -o report -yf 2016 -mf 05 -yt 2016 -mt 06

Sifting documents between 01-May-2016 and 30-Jun-2016
22 documents available after publication date sift
Dropped 0 patents due to empty abstracts
Processing TFIDF of 22 documents
No focus applied
1. shape surface                  0.516398
2. main compartment               0.478091

### CODE REVIEW

File: pygrams.py [hunk 1]
```diff
@@ -49,8 +64,10 @@ def get_args(command_line_arguments):
                         help="options are: <fdg> <wordcloud> <report> <table> <tfidf> <all>")
     parser.add_argument("-j", "--json", default=False, action="store_true",
                         help="Output configuration as JSON file alongside output report")
-    parser.add_argument("-yf", "--year_from", type=int, default=2000, help="The first year for the document cohort")
-    parser.add_argument("-yt", "--year_to", type=int, default=0, help="The last year for the documents cohort (0 is now)")
+    parser.add_argument("-yf", "--year_from", default=None, help="The first year for the document cohort in YYYY format")
+    parser.add_argument("-mf", "--month_from", default=None, help="The first month for the document cohort in MM format")
+    parser.add_argument("-yt", "--year_to", default=None, help="The last year for the documents cohort in YYYY format")
```
Comment (IanGrimstead) [inline pygrams.py]:
```suggestion
    parser.add_argument("-yt", "--year_to", default=None, help="The last year for the document cohort in YYYY format")
```

File: pygrams.py [hunk 2]
```diff
@@ -49,8 +64,10 @@ def get_args(command_line_arguments):
...
+    parser.add_argument("-mt", "--month_to", default=None, help="The last month for the documents cohort in MM format")
```
Comment (IanGrimstead) [inline pygrams.py]:
```suggestion
    parser.add_argument("-mt", "--month_to", default=None, help="The last month for the document cohort in MM format")
```

File: pygrams.py [hunk 3]
```diff
@@ -83,9 +100,29 @@ def get_args(command_line_arguments):
 
 def checkargs(args):
     app_exit = False
-    if args.year_to != 0:
-        if args.year_from >= args.year_to:
-            print("year_from must be less than year_to")
+    if isinstance(args.year_to,str) & isinstance(args.year_from,str):
```
Comment (IanGrimstead) [inline pygrams.py]: Months may not be defined when years are mandatory? So this may blow if months aren't defined?

Comment (mshodge) [reply pygrams.py]: Resolved in the latest commits I believe. year_to/year_from are needed if month_to/month_from are entered as well now. Should resolve all issues I hope.

**Comment by mshodge** [comment]
@IanGrimstead just added more arg checks so should be ok now 😄

**Comment by IanGrimstead** [comment]
Only thought is - any impact on JSON format changing?

**Comment by IanGrimstead** [comment]
If you're after a challenge - we should have tested the args code so we can prove all the edge cases are caught :) Allegedly I think you've got them tho'

**Comment by mshodge** [comment]
JSON is fine, had to write some separate if else statements for it but it outputs as expected 😄 

Could write some args testing next week if we want.

**Comment by mshodge** [comment]
@IanGrimstead detect.py updated now. Also caught a bug (doc_set) on that too which is resolved now.

**Comment by mshodge** [comment]
Reviewing the documentation for PyGrams and was wondering if this could get accepted?

**Comment by IanGrimstead** [comment]
Unblocked – you can merge now at your convenience 😊

**Pull request merged**